server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregator.java

```diff
@@ -0,0 +1,157 @@
+
+        super(name, context, parent, pipelineAggregators, metaData);
+
+        this.valuesSource = valuesSource;
+        this.format = format;
+
+        this.valueSketches = context.bigArrays().newObjectArray(1);
+        this.compression = compression;
+    }
+
+    @Override
+    public double metric(long owningBucketOrd) {
+        if (owningBucketOrd >= valueSketches.size() || valueSketches.get(owningBucketOrd) == null) {
+            return Double.NaN;
+        } else {
+            return computeMedianAbsoluteDeviation(valueSketches.get(owningBucketOrd));
+        }
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        if (valuesSource != null && valuesSource.needsScores()) {
+            return ScoreMode.COMPLETE;
+        } else {
+            return ScoreMode.COMPLETE_NO_SCORES;
+        }
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (valuesSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+
+        final BigArrays bigArrays = context.bigArrays();
+        final SortedNumericDoubleValues values = valuesSource.doubleValues(ctx);
+
+        return new LeafBucketCollectorBase(sub, values) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+
+                valueSketches = bigArrays.grow(valueSketches, bucket + 1);
+
+                TDigestState valueSketch = valueSketches.get(bucket);
+                if (valueSketch == null) {
+                    valueSketch = new TDigestState(compression);
+                    valueSketches.set(bucket, valueSketch);
+                }
+
+                if (values.advanceExact(doc)) {
+                    final int valueCount = values.docValueCount();
+                    for (int i = 0; i < valueCount; i++) {
+                        final double value = values.nextValue();
+                        valueSketch.add(value);
+                    }
+                }
+            }
+        };
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long bucket) throws IOException {
+        if (bucket >= valueSketches.size() || valueSketches.get(bucket) == null) {
+            return buildEmptyAggregation();
+        } else {
+            final TDigestState valueSketch = valueSketches.get(bucket);
+            return new InternalMedianAbsoluteDeviation(name, pipelineAggregators(), metaData(), format, valueSketch);
+        }
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalMedianAbsoluteDeviation(name, pipelineAggregators(), metaData(), format, new TDigestState(compression));
+    }
+
+    @Override
+    public void doClose() {
+        Releasables.close(valueSketches);
+    }
+
+    public static double computeMedianAbsoluteDeviation(TDigestState valuesSketch) {
+
+        if (valuesSketch.size() == 0) {
```

Comment:
Since unmapped aggs have a `null` sketch I think we need a null check in this method?

Reply:
I assume the unmapped and partially unmapped tests in the integration test fail at the moment because of this? If not, we should work out why and ensure they are correct and have a test that tests for NPEs here

Reply:
This method actually has the same behavior (returning NaN) without this check, because we create an empty `approximatedDeviationsSketch`, and calling `#quantile` on an empty sketch returns NaN. I added this check to make it really obvious what happens without data points, since I don't think that `#quantile` returning NaN is documented

Reply:
I explained why I don't think `valuesSketch` can be null here in another comment, the unmapped and partially unmapped tests were passing when I committed this